soroban-sdk/src/crypto/bls12_381.rs

```diff
@@ -0,0 +1,294 @@
+/// - The 48 bytes represent the **big-endian encoding** of an element in the
+///   field `Fp`. The value is serialized as a big-endian integer.
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct Fp(BytesN<48>);
+
+/// # `Fp2` represents an element of the quadratic extension field `Fq2` of the
+/// BLS12-381 elliptic curve
+///
+/// # Serialization:
+/// - The 96 bytes represent the **big-endian encoding** of an element in the
+///   field `Fp2`. The bytes consist of `be_bytes(c1) || be_bytes(c0)` (`||` is
+///   concatenation), where `c0` and `c1` are the two `Fp` elements (the real
+///   and imaginary components).
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct Fp2(BytesN<96>);
+
+impl_bytesn_repr!(G1Affine, 96);
+impl_bytesn_repr!(G2Affine, 192);
+impl_bytesn_repr!(Fp, 48);
+impl_bytesn_repr!(Fp2, 96);
+
+impl Bls12_381 {
+    pub(crate) fn new(env: &Env) -> Bls12_381 {
+        Bls12_381 { env: env.clone() }
+    }
+
+    pub fn env(&self) -> &Env {
+        &self.env
+    }
+
+    // g1
+
+    /// Checks if a point `p` in G1 is in the correct subgroup.
+    pub fn g1_is_in_subgroup(&self, p: &G1Affine) -> bool {
+        let env = self.env();
+        let res = internal::Env::bls12_381_check_g1_is_in_subgroup(env, p.to_object())
+            .unwrap_infallible();
+        res.into()
+    }
+
+    /// Adds two points `p0` and `p1` in G1.
+    pub fn g1_add(&self, p0: &G1Affine, p1: &G1Affine) -> G1Affine {
+        let env = self.env();
+        let bin = internal::Env::bls12_381_g1_add(env, p0.to_object(), p1.to_object())
+            .unwrap_infallible();
+        G1Affine::from_bytes(bin.into_val(env))
+    }
+
+    /// Adds two points `p0` and `p1` in G1, ensuring that the result is in the
+    /// correct subgroup. Note the subgroup check is computationally expensive,
+    /// so if want to perform a series of additions i.e. `agg = p0 + p1 + .. + pn`,
+    /// it may make sense to only call g1_checked_add on the final addition,
+    /// while using `g1_add` (non-checked version) on the intermediate ones.
+    pub fn g1_checked_add(&self, p0: &G1Affine, p1: &G1Affine) -> Option<G1Affine> {
+        let env = self.env();
+        let bin = internal::Env::bls12_381_g1_add(env, p0.to_object(), p1.to_object())
+            .unwrap_infallible();
+        let res = G1Affine::from_bytes(bin.into_val(env));
+        let is_in_correct_subgroup: bool =
+            internal::Env::bls12_381_check_g1_is_in_subgroup(env, res.to_object())
+                .unwrap_optimized()
+                .into();
+        match is_in_correct_subgroup {
+            true => Some(res),
+            false => None,
+        }
+    }
+
+    /// Multiplies a point `p0` in G1 by a scalar.
+    pub fn g1_mul(&self, p0: &G1Affine, scalar: &U256) -> G1Affine {
+        let env = self.env();
+        let bin =
+            internal::Env::bls12_381_g1_mul(env, p0.to_object(), scalar.into()).unwrap_infallible();
+        G1Affine::from_bytes(bin.into_val(env))
```

Comment:
This line is doing unnecessary checks and adding unnecessary branches and panics into the code. This into_val is doing a check on the bin/bytes size when converting the BytesObject into a BytesN, and then a branch to panic if it's not the right size.

We should I think use the unchecked_new function, since the env guarantees the size of the bytes object returned:
https://github.com/stellar/rs-soroban-sdk/blob/1367be1ad293e8b75cb6de5814a88391a69562b8/soroban-sdk/src/bytes.rs#L984-L988

Reply:
Done. I updated every occurrence of this issue.